File: typescript/6.ts
function convert(s: string, numRows: number): string {
    if (numRows === 1) return s;

    const m: Map<number, string> = new Map();
    let vertical = true;
    let line = 0;
    for (let char of s) {
        m.set(line, (m.get(line) || '') + char);
        line += vertical ? 1 : -1;

        if (line === numRows - 1) vertical = false;
        else if (line === 0) vertical = true;
    }

    let ret = '';
    for (let i = 0; i < numRows; i++) ret += m.get(i) || '';
    return ret;
};

test('6', () => {
    expect(convert("PAYPALISHIRING", 3)).toBe("PAHNAPLSIIGYIR");
});

File: python/6.py
import unittest
from typing import Dict


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        m: Dict[int, str] = {}
        vertical = True
        line = 0
        for char in s:
            m[line] = m.get(line, '') + char
            line += 1 if vertical else -1

            if line == numRows - 1:
                vertical = False
            elif line == 0:
                vertical = True

        ret = ''
        for i in range(numRows):
            ret += m.get(i, '')
        return ret


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.convert(
            'PAYPALISHIRING', 3), 'PAHNAPLSIIGYIR')
